pallets/foreign-investments/src/tests.rs

```diff
@@ -904,33 +981,95 @@ mod investment {
 			util::fulfill_last_swap(Action::Investment, AMOUNT);
 
 			let total_foreign_collected = Arc::new(Mutex::new(0));
-			let foreign_remaining = Arc::new(Mutex::new(0));
 
 			for _ in 0..foreign_to_pool(AMOUNT) {
 				util::process_investment(1 /* pool_amount */);
 
-				MockCollectInvestHook::mock_notify_status_change({
+				MockHooks::mock_fulfill_collect_investment({
 					let total_foreign_collected = total_foreign_collected.clone();
-					let foreign_remaining = foreign_remaining.clone();
-					move |_, msg| {
+					move |_, _, _, amount_collected, _| {
 						// First messages returns nothing, until last messages fix the expected
 						// returned value.
 
-						*total_foreign_collected.lock().unwrap() += msg.amount_currency_payout;
-						*foreign_remaining.lock().unwrap() = msg.amount_remaining;
+						*total_foreign_collected.lock().unwrap() += amount_collected;
 						Ok(())
 					}
 				});
 
-				assert_ok!(ForeignInvestment::collect_foreign_investment(
-					&USER,
-					INVESTMENT_ID,
-					FOREIGN_CURR
-				));
+				assert_ok!(MockInvestment::collect_investment(USER, INVESTMENT_ID));
 			}
 
 			assert_eq!(*total_foreign_collected.lock().unwrap(), AMOUNT);
-			assert_eq!(*foreign_remaining.lock().unwrap(), 0);
+		});
+	}
+
+	#[test]
+	fn increase_and_fulfill_and_very_small_partial_collects_and_cancel() {
+		// Rate is: 1 pool amount = 0.1 foreign amount.
+		// There is no equivalent foreign amount to return when it collects just 1 pool
+		// token, so most of the first messages seems to return nothing.
+		//
+		// Nevertheless the system can recover itself from this situation and the
+		// accumulated result is the expected one.
+		new_test_ext().execute_with(|| {
+			util::base_configuration();
+
+			assert_ok!(ForeignInvestment::increase_foreign_investment(
+				&USER,
+				INVESTMENT_ID,
+				AMOUNT,
+				FOREIGN_CURR
+			));
+
+			util::fulfill_last_swap(Action::Investment, AMOUNT);
+
+			let foreign_fulfilled = Arc::new(Mutex::new(0));
```

Comment:
Curious question: It's not really necessary to do this over using `mut` right because we are not leveraging concurrency yet in UTs?

Reply:
Theoretically, there are not any concurrency issues, but the `mock-builder` expects any moved reference to the closure to be `'static`, so we can not pass a `&mut` reference to mutate it inside the closure. Also, `mock-builder` is not thread-safe (although theoretically, it could be for the same tests), so the closure can not be `Sync` and `Send`. So we need to pass an `Arc` (`Rc` is not supported)